Bugzilla/DaemonControl.pm

```diff
@@ -0,0 +1,317 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::DaemonControl;
+use 5.10.1;
+use strict;
+use warnings;
+
+use Cwd qw(realpath);
+use English qw(-no_match_vars $PROGRAM_NAME);
+use File::Spec::Functions qw(catfile);
+use Future::Utils qw(repeat try_repeat);
+use Future;
+use IO::Async::Loop;
+use IO::Async::Process;
+use IO::Async::Protocol::LineStream;
+use IO::Async::Signal;
+use LWP::Simple qw(get);
+use POSIX qw(setsid WEXITSTATUS);
+use WWW::Selenium::Util qw(server_is_running);
+use Bugzilla::Constants qw(bz_locations);
+
+use base qw(Exporter);
+
+our @EXPORT_OK = qw(
+    run_httpd run_cereal run_cereal_and_httpd
+    catch_signal on_finish on_exception
+    assert_httpd assert_database assert_selenium
+);
+
+our %EXPORT_TAGS = (
+    all => \@EXPORT_OK,
+    run => [grep { /^run_/ } @EXPORT_OK],
+    utils => [qw(catch_signal on_exception on_finish)],
+);
+
+use constant HTTPD_BIN     => '/usr/sbin/httpd';
+use constant HTTPD_CONFIG  => realpath(catfile( bz_locations->{confdir}, 'httpd.conf' ));
+
+sub catch_signal {
+    my ($name, @done)   = @_;
+    my $loop     = IO::Async::Loop->new;
+    my $signal_f = $loop->new_future;
+    my $signal   = IO::Async::Signal->new(
+        name       => $name,
+        on_receipt => sub {
+            $signal_f->done(@done);
+        }
+    );
+    $signal_f->on_cancel(
+        sub {
+            my $l = IO::Async::Loop->new;
+            $l->remove($signal);
+        },
+    );
+
+    $loop->add($signal);
+
+    return $signal_f;
+}
+
+sub cereal {
+    local $PROGRAM_NAME = "cereal";
+    my $loop = IO::Async::Loop->new;
+    $loop->listen(
+        host => '127.0.0.1',
+        service => '5880',
+        socktype => 'stream',
+        on_stream => sub {
+            my ($stream) = @_;
+            my $protocol = IO::Async::Protocol::LineStream->new(
+                transport => $stream,
+                on_read_line => sub {
+                    my ($self, $line) = @_;
+                    say $line;
+                },
+            );
+            $loop->add($protocol);
+        },
+    )->get;
+    kill 'USR1', getppid();
+
+    exit catch_signal('TERM', 0)->get;
+}
+
+sub run_cereal {
+    my $loop   = IO::Async::Loop->new;
+    my $exit_f = $loop->new_future;
+    my $cereal = IO::Async::Process->new(
+        code         => \&cereal,
+        on_finish    => on_finish($exit_f),
+        on_exception => on_exception( "cereal", $exit_f ),
+    );
+    $exit_f->on_cancel( sub { $cereal->kill('TERM') } );
+    $loop->add($cereal);
+    catch_signal('USR1')->get;
```

Comment:
is there a small race condition here?
if the USR1 signal is sent before the signal handler is setup this would kill the process

Reply:
I'm not sure if the process starts at that time, but there is a clear way to rewrite this with futures, so I'll do this:

```perl
my $signal_f = catch_signal('USR1');
 $loop->add($cereal);
$signal_f->get;
```